From 415b992217a1e393fac5f4a6893013135510cda0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20P=C3=ADsa=C5=99?= <ppisar@redhat.com>
Date: Tue, 17 Jan 2017 13:48:37 +0100
Subject: [PATCH] Fix warning on printing pointer difference
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 src/pcre2test.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/pcre2test.c b/src/pcre2test.c
index 5ea245d..e1c8d25 100644
--- a/src/pcre2test.c
+++ b/src/pcre2test.c
@@ -4663,7 +4663,7 @@ if ((pat_patctl.control & CTL_HEXPAT) != 0)
         if (d == 0)
           {
           fprintf(outfile, "** Missing closing quote in hex pattern: "
-            "opening quote is at offset %lu.\n", pq - buffer - 2);
+            "opening quote is at offset %td.\n", pq - buffer - 2);
           return PR_SKIP;
           }
         if (d == c) break;
@@ -4677,7 +4677,7 @@ if ((pat_patctl.control & CTL_HEXPAT) != 0)
       {
       if (!isxdigit(c))
         {
-        fprintf(outfile, "** Unexpected non-hex-digit '%c' at offset %lu "
+        fprintf(outfile, "** Unexpected non-hex-digit '%c' at offset %td "
           "in hex pattern: quote missing?\n", c, pp - buffer - 2);
         return PR_SKIP;
         }
@@ -4689,7 +4689,7 @@ if ((pat_patctl.control & CTL_HEXPAT) != 0)
       d = *pp;
       if (!isxdigit(d))
         {
-        fprintf(outfile, "** Unexpected non-hex-digit '%c' at offset %lu "
+        fprintf(outfile, "** Unexpected non-hex-digit '%c' at offset %td "
           "in hex pattern: quote missing?\n", d, pp - buffer - 1);
         return PR_SKIP;
         }
-- 
2.7.4

